Added support for federated authentication to enable Azure AD authentication

wrosenuance opened the pull request:
This is a pull request to add support for the federated authentication and security token library login options. Federated authentication is used for Active Directory logins in Azure, particularly for user/password and managed identity logins, while the security token library handles AD application (service principal) authentication with client secrets or client certificates.

The PR includes a Terraform environment setup that provisions a VM and Azure SQL server with the appropriate role grants to verify that managed identity login is working, which is necessary to execute the full suite of tests. Tests that rely on Azure VM infrastructure are disabled if the environment settings do not provide the right DSN, to avoid impact to existing testing.

wrosenuance commented:
The low code coverage reported from AppVeyor is because I cannot enable AD logins and MSI features against their SQL server. Testing these features to get better coverage can be done manually - I have a guide at https://github.com/wrosenuance/go-mssqldb/blob/azure-auth/doc/how-to-test-azure-ad-authentication.md. If you would like automated testing for coverage I think this will require adding a mocking library (like counterfeiter) and revising much more of the structure to support injecting mocks.

kardianos commented:
@wrosenuance I just sent a review #546 . I'm going to wait for that PR to merge before doing a review on this one, but doing a skim on this one I would request the following general changes:

 * If possible, inject the adal functionality if possible (so this module doesn't direclty depend on it. I like how #546 does it. If not possible, that might be fine, but I'll need to understand why. It looks like `adal` package is only used in a single function, so it may be possible to inject that in the connector or similar.
 * Please replace if/else chains with either early returns or switch statements.
 * We will want to add a go.mod file under the example folder, #546 may do that.

wrosenuance commented:
Thanks @kardianos! I think @paulmey had reviewed this PR and had offered to rebase his contribution on top of this one, so that #546 is replaced by #548, with the expectation that it would merge after this one. Is it possible to apply them in that order instead?

wrosenuance commented:
> * If possible, inject the adal functionality if possible (so this module doesn't direclty depend on it. I like how #546 does it. If not possible, that might be fine, but I'll need to understand why. It looks like `adal` package is only used in a single function, so it may be possible to inject that in the connector or similar.

The key reason for integrating them is user convenience - if the modules are not integrated, driver users will have to call driver-specific constructors and possibly write code to wire in AD authentication, whereas the PR as proposed makes this configurable via connection string parameters that are available generically to users working through the driver-agnostic sql package.

> * Please replace if/else chains with either early returns or switch statements.

I will look into this - I think you pointed out a few in your review of #546 that are pertinent here as well.

> * We will want to add a go.mod file under the example folder, #546 may do that.

Ok I will check!

kardianos commented:
I've merged #546, please rebase on master and I'll review this closely.

kardianos commented on tds.go:
```diff
@@ -835,23 +1021,54 @@ initiate_connection:
 	}
 	auth, auth_ok := getAuth(p.user, p.password, p.serverSPN, p.workstation)
-	if auth_ok {
+	if login.FedAuthLibrary == fedAuthLibraryReserved && auth_ok {
```
This section +30 lines of the nested else/if needs to be cleaned up. using switch may help, and maybe try to get rid of nested ifs with early returns.

wrosenuance replied:
I have restructured this as a switch.

kardianos commented on fedauth.go:
```diff
@@ -0,0 +1,101 @@
+	return certificate, rsaPrivateKey, nil
+}
+
+func fedAuthGetAccessToken(ctx context.Context, resource, tenantID string, p connectParams, log optionalLogger) (accessToken string, err error) {
```
I'd really like to see this function isolated in a sub package at a minimum. Maybe we can still directly reference it, but I would really like it injected somehow.

wrosenuance replied:
I could try to refactor this so that it uses a package initialization routine to add in the Azure AD support? Would that work? Then a user that wanted Azure AD via connection string parameters would need to use two imports:
```go
import (
	"database/sql"

	_ "github.com/denisenkom/go-mssqldb"
	_ "github.com/denisenkom/go-mssqldb/azure"
)
```
If that's acceptable, I think it would be possible to make the direct dependency opt-in like that.

wrosenuance replied:
Actual proposal has the import path as github.com/denisenkom/go-mssqldb/azuread, and it is sufficient to just import that package if the generic sql.Open method is used.

kardianos commented on fedauth.go:
```diff
@@ -0,0 +1,101 @@
+	// URL for the tenant, the URL resolution just returns that endpoint.
+	oauthConfig, err := adal.NewOAuthConfig(activeDirectoryEndpoint, tenantID)
+	if err != nil {
+		log.Printf("Failed to obtain OAuth configuration for endpoint %s and tenant %s: %v", activeDirectoryEndpoint, tenantID, err)
```
You must not log like this.
Rather, you can do return "", fmt.Errorf("Failed to obtain... %v", err)
but no log to console.

wrosenuance replied:
The call to log.Printf was being directed to the driver logger - not the console, but I changed it to a wrapped error return.

kardianos commented on fedauth.go:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	var token *adal.ServicePrincipalToken
+	if p.fedAuthLibrary == fedAuthLibrarySecurityToken {
```
Try to get rid of these if/else sequences and nested if/elses. Use switch and early returns or initial value and overrdie.

wrosenuance replied:
These were also revised to use a switch.

wrosenuance commented:
Hi @kardianos and @paulmey! I pushed an update to the branch to rebase and merge with the PR #546 from @paulmey. 

I have tried to incorporate the requests above in terms of restructuring code flow and moving the ADAL code to a separate package. There are some drawbacks to this - even though it's possible to just switch the import to re-enable the ADAL authentication, it seems clumsy, and I can no longer run the existing test suite in the top-level package against an Azure SQL database authenticated via Active Directory.

As this PR adds support for the ADAL workflow where the server provides the resource and STS URL, there needs to be a different function signature for providing tokens that accepts these, plus I think the context ought to be passed in to the existing access token connector. Since there was already an option to provide a custom Dialer in the existing Connector structure, I moved the security token and Active Directory providers there.

Please let me know what you think of the revised code and approach - particularly whether it is okay to have the package separation as proposed here.

kardianos commented:
@wrosenuance Thank you for the updates. There are a number of changes going on and it will take me a few days to review. Thank you for being patient with me and my requests so far.

wrosenuance commented:
@kardianos, did you get a chance to review this yet? Please let me know if there is any way to help!

serbrech commented:
Hi, anyone is still working on this?

wrosenuance commented:
@serbrech I am still using my branch in my day job, I'm still waiting for a review for this. @kardianos - is this something you can pick up again?

kardianos commented:
Sorry for the delay. This year has been a blast. I've stared this change, will try get to it.

serbrech commented:
if it's easier, could we reduce the scope of the PR to add the different options one by one?
for example, I'm only interesting in the MSI one. Happy to help land it if needed.

wrosenuance commented:
I'm not sure it gets much easier: the main protocol changes are required either way, the exact method for obtaining credentials is a little more removed from the core of the change.

serbrech commented on README.md:
```diff
@@ -54,10 +54,15 @@ Other supported formats are listed below.
-
+* `FedAuth` - The federated authentication scheme to use. See below for additional setup requirements.
+  * `ActiveDirectoryApplication` - authenticates using an Azure Active Directory application client ID and client secret or certificate. Set the `user` to `client-ID@tenant-ID` and the `password` to the client secret. If using client certificates, provide the path to the PKCS#12 file containing the certificate and RSA private key in the `ClientCertPath` parameter, and set the `password` to the va ...
+  * `ActiveDirectoryMSI` - authenticates using the managed service identity (MSI) attached to the VM, or a specific user-assigned identity if a client ID is specified in the `user` field.
```
next step wish : 
avoid specifying the clientID.
useMSI + identityName or resourceID would be nicer, as the clientID is only known after deploying the MSI. 
the resourceID/name is known before deployment. this helps in configuring multiple environment statically.

wrosenuance replied:
I think to look up the clientID you would need Azure resource manager (portal) access, or at least Azure AD, and I think you might need an initial authentication context to be able to do that. So you would have a chicken-and-egg situation - to get the client ID to use as part of your credentials, you'd need credentials. You may have more luck working on a deploy-time solution to look up the client ID though.

When I've been deploying this I was using the Kubernetes AAD pod identity controller, and the client ID is put in there. We deploy using Azure DevOps pipelines, and the deploy pipeline is able to do the lookup using credentials available to the pipeline.

serbrech replied:
no, the IMDS endpoint on the host has an endpoint to get the token from the MSI resource ID. we added support for that in ADAL just a few weeks ago : 
https://github.com/Azure/go-autorest/pull/544

serbrech replied:
"When I've been deploying this I was using the Kubernetes AAD pod identity controller, and the client ID is put in there. We deploy using Azure DevOps pipelines, and the deploy pipeline is able to do the lookup using credentials available to the pipeline."
that means you have dynamic configuration, fetching an ID at deployment time.
I would rather be able to avoid this, and keep the configuration static :). 
not a necessity for this PR, just providing my point of view based on real case constraints we have

wrosenuance replied:
Oh nice! I had not seen that!

wrosenuance replied:
LOL but I see now the PR is failing in CI because of other changes in the go-autorest library: we may not be able to use the more recent library and maintain the Go version compatibility (1.8+) that was previously possible.

serbrech replied:
I'll send a small PR to just add the context to the tokenprovider func.
that would make the current ode viable, while waiting for this to land.
Thanks a lot for this work though. again, happy to help land it if needed

wrosenuance replied:
I put in a workaround in the CI builds to check out the last tag that builds from the ADAL project - I think it actually includes the change you've mentioned above, though.

wrosenuance replied:
And I created [an issue](https://github.com/Azure/go-autorest/issues/555) in the go-autorest project: I'm not sure if they consider supporting Go prior to 1.13 important, though.

serbrech replied:
the issue mentioned above is fixed in latest version of autorest. it should be good to go.

serbrech commented on azuread/fedauth_adal.go:
```diff
@@ -0,0 +1,136 @@
+)
+
+func retrieveToken(ctx context.Context, token *adal.ServicePrincipalToken) (string, error) {
+	err := token.RefreshWithContext(ctx)
```
should this rather use `token.EnsureFreshWithContext(ctx)`?

wrosenuance replied:
Possibly! I don't recall deciding not to use that, I think the RefreshWithContext call was part of an ADAL code sample that I just used as-is.

serbrech commented on mssql.go:
```diff
@@ -91,6 +92,21 @@ func NewConnector(dsn string) (*Connector, error) {
+// security token is required.  The string returned should be the access
+// token to supply to the server, otherwise an error can be returned to
+// indicate why a token is not available.
+type SecurityTokenProvider func(ctx context.Context) (string, error)
```
Ah! I was looking at where you changed the provider func to take in the context.
nicer. thanks.

serbrech commented on azuread/fedauth_adal.go:
```diff
@@ -0,0 +1,136 @@
+
+	// azureSQLResource is the AD resource to use when the server does not
+	// provide the resource.
+	azureSQLResource = "https://database.windows.net/"
```
this needs to be configurable.
not all azure cloud have the same database resource endpoint.
I would say it's a good default, but it would be nice to be able to modify it via an environment variable, for example

serbrech replied:
use this func instead to get the correct endpoint : 

https://github.com/Azure/go-autorest/blob/1d7168cf650ab7fdc4e05297b7b0250be08d0a8a/autorest/azure/environments.go#L227

`SQLDatabaseDNSSuffix` has the correct url on the environment struct

serbrech replied:
or better yet, infer it from the connection string?

arvindshmicrosoft commented:
Hello @wrosenuance - thanks for the great work on this. We have a use case where the changes enabled by this PR would greatly simplify the handling of Managed Identity (MSI) based authentication for the [Telegraf sqlserver plugin](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/sqlserver/sqlserver.go). Do you have any plans to address the open comments? Is there anything I can do to help?

wrosenuance commented:
> Hello @wrosenuance - thanks for the great work on this. We have a use case where the changes enabled by this PR would greatly simplify the handling of Managed Identity (MSI) based authentication for the Telegraf sqlserver plugin

Awesome!!

> Do you have any plans to address the open comments? Is there anything I can do to help?

I have been waiting for @kardianos to take a look - after his initial review I made some changes but I have been holding off making many more as I'm waiting to get an indication of a path to merging this. For instance, if it made sense to split this up, it might have have a big effect on the way the PR proceeds.

wrosenuance commented:
Oh! I see some recent changes from @denisenkom have introduced conflicts so I should also adapt the PR to match. 😅

wrosenuance commented:
Hi @denisenkom and @kardianos, I rebased this PR to work again after the recent commits from @denisenkom and added some more test cases to increase the coverage. Could you please review and let me know if anything else is needed for this to be merged?

kardianos commented:
@wrosenuance Thank you for doing this. I just went through this an initial time in this form.

This change is large. Would it be possible for you to open a new PR with changes just to the mssql package, with as few changes as possible to existing code?

wrosenuance commented:
Thanks @kardianos! I will split it into a few PRs and link them here. I am thinking I can make the following large chunks:

- add the connection logger code that is used to dump the packet-level activity
- add the new token types and structure definitions - I think these can exist before being used
- add the main functional changes to the mssql package
- add the ADAL token routines from the azuread package
- add the documentation

Please let me know if that doesn't work!

kardianos commented:
If you could start with just the changes to mssql package, both tokens and function changes, that would be great.

shueybubbles commented:
I think we should revise this plan a bit and move to 	"github.com/Azure/azure-sdk-for-go/sdk/azidentity" 
It provides a broad set of credential types with little extra code needed in the driver.
Here's a piece of code from an app I'm building which uses `azidentity` on top of `go-mssqldb`
The big missing piece for me is the lack of server-provided tenant and resource url.
Also, we need to support an application-provided client id.

```go

func (s *Sqlcmd) GetTokenBasedConnection(connstr string, user string, password string) (driver.Connector, error) {
	var cred azcore.TokenCredential
	var err error
	scope := ".default"
	t := azureTenantId()
	switch s.Connect.AuthenticationMethod {
	case ActiveDirectoryDefault:
		cred, err = azidentity.NewDefaultAzureCredential(nil)
	case ActiveDirectoryInteractive:
		cred, err = azidentity.NewInteractiveBrowserCredential(&azidentity.InteractiveBrowserCredentialOptions{TenantID: t, ClientID: getSqlClientId()})
// this scope may not be needed
		scope = "user_impersonation"
	case ActiveDirectoryPassword:
		cred, err = azidentity.NewUsernamePasswordCredential(t, getSqlClientId(), user, password, nil)
	case ActiveDirectoryManagedIdentity:
		cred, err = azidentity.NewManagedIdentityCredential(user, nil)
	case ActiveDirectoryServicePrincipal:
		cred, err = azidentity.NewClientSecretCredential(t, user, password, nil)
	default:
		// no implementation of AAD Integrated yet
		cred, err = azidentity.NewDefaultAzureCredential(nil)
	}

	if err != nil {
		return nil, err
	}
	resourceUrl := s.getResourceUrl()
	conn, err := mssql.NewAccessTokenConnector(connstr, func() (string, error) {
		opts := policy.TokenRequestOptions{Scopes: []string{resourceUrl + scope}}
		tk, err := cred.GetToken(context.Background(), opts)
		if err != nil {
			return "", err
		}
		return tk.Token, err
	})

	return conn, err
}

```
